main.rb

```diff
@@ -0,0 +1,172 @@
+module Enumerable
+  def my_each
+    pos = 0
+    loop do
+      break if pos >= length
+
+      yield self[pos]
+      pos += 1
+    end
+  end
+
+  def my_each_with_index
+    pos = 0
+    loop do
+      break if pos >= length
+
+      yield self[pos], pos
+      pos += 1
+    end
+  end
+
+  def my_select
+    return_arr = []
+    my_each do |element|
+      return_arr.push(element) if yield element
+    end
+    return_arr
+  end
+
+  def my_all
+    return_val = true
+    my_each do |element|
+      return_val = yield element
+    end
+    return_val
+  end
+
+  def my_any
+    return_val = false
+    my_each do |element|
+      return_val = yield element
+    end
+    return_val
+  end
+
+  def my_none
+    return_val = true
+    my_each do |element|
+      return_val = !(yield element)
+    end
+    return_val
+  end
+
+  def my_count(arg = nil)
+    count_val = 0
+    my_each do |element|
+      count_val += 1 if (block_given? && yield(element)) || (element == arg)
+    end
+    count_val
+  end
+
+  def my_map(proc_arg = nil)
+    return_arr = []
+    my_each_with_index do |element, index|
+      if proc_arg
+        return_arr.push(proc_arg.call(element))
+      elsif block_given?
+        return_arr.push(yield(element))
+      else
+        return_arr.push(index)
+      end
+    end
+    return_arr
+  end
+
+  def my_inject(*args)
+    memo = args[0].is_a?(Symbol) ? self[0] : args[0] || self[0]
+    symbol = identify_symbol(*args)
+    pos = settle_start_position(*args)
+    while pos < length
+      memo = block_given? ? yield(memo, self[pos]) : memo.send(symbol, self[pos])
+      pos += 1
+    end
+    memo
+  end
+end
```

Comment:
* [ ] `my_inject` is identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(4) { |prod, n| prod * n } # returns this error
    #NameError:
        undefined local variable or method `length' for 5..50:Range
      # ./main.rb:80:in `my_inject'
  ```

* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(2, :*) # returns this error
    #  NameError:
        undefined local variable or method `length' for 5..50:Range
      # ./main.rb:80:in `my_inject'
  ```